File: python-backend/app/routers/credits.py
"""
SmartSpec Pro - Credits Router
API endpoints for credits management and subscription plans.
"""

from datetime import datetime
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Request, status
from pydantic import BaseModel, Field
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.auth import get_current_user
from app.models.user import User
from app.models.credits import TransactionType, UsageType
from app.services.generation.credits_repository import (
    CreditsRepository,
    get_credits_repository,
)

router = APIRouter(prefix="/credits", tags=["Credits"])


# =============================================================================
# REQUEST/RESPONSE SCHEMAS
# =============================================================================

class BalanceResponse(BaseModel):
    """Credits balance response."""
    total_credits: float
    used_credits: float
    reserved_credits: float
    available_credits: float
    subscription_tier: str
    monthly_allowance: float


class TransactionResponse(BaseModel):
    """Credit transaction response."""
    id: str
    transaction_type: str
    amount: float
    balance_before: float
    balance_after: float
    reference_type: Optional[str]
    reference_id: Optional[str]
    description: Optional[str]
    created_at: datetime


class TransactionListResponse(BaseModel):
    """List of transactions."""
    transactions: List[TransactionResponse]
    count: int


class UsageSummaryResponse(BaseModel):
    """Usage summary response."""
    summary: dict
    period_start: Optional[datetime]
    period_end: Optional[datetime]


class DailyUsageResponse(BaseModel):
    """Daily usage response."""
    data: List[dict]
    days: int


class ModelUsageResponse(BaseModel):
    """Model usage statistics response."""
    data: List[dict]
    days: int


class PlanResponse(BaseModel):
    """Subscription plan response."""
    id: str
    name: str
    display_name: str
    description: Optional[str]
    price_monthly: float
    price_yearly: float
    currency: str
    monthly_credits: float
    bonus_credits: float
    max_storage_mb: int
    features: List[str]


class PlansListResponse(BaseModel):
    """List of subscription plans."""
    plans: List[PlanResponse]


class AddCreditsRequest(BaseModel):
    """Request to add credits (admin only)."""
    amount: float = Field(..., gt=0, description="Amount of credits to add")
    reason: str = Field(..., min_length=1, description="Reason for adding credits")


# =============================================================================
# BALANCE ENDPOINTS
# =============================================================================

@router.get("/balance", response_model=BalanceResponse)
async def get_balance(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Get current user's credits balance."""
    repo = get_credits_repository(db)
    balance = await repo.get_or_create_balance(current_user.id)
    
    return BalanceResponse(
        total_credits=balance.total_credits,
        used_credits=balance.used_credits,
        reserved_credits=balance.reserved_credits,
        available_credits=balance.available_credits,
        subscription_tier=balance.subscription_tier,
        monthly_allowance=balance.monthly_allowance,
    )


# =============================================================================
# TRANSACTION ENDPOINTS
# =============================================================================

@router.get("/transactions", response_model=TransactionListResponse)
async def get_transactions(
    transaction_type: Optional[str] = None,
    limit: int = 50,
    offset: int = 0,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Get credit transaction history."""
    repo = get_credits_repository(db)
    
    type_enum = TransactionType(transaction_type) if transaction_type else None
    
    transactions = await repo.get_transactions(
        user_id=current_user.id,
        transaction_type=type_enum,
        limit=min(limit, 100),
        offset=offset,
    )
    
    return TransactionListResponse(
        transactions=[
            TransactionResponse(
                id=str(t.id),
                transaction_type=t.transaction_type.value,
                amount=t.amount,
                balance_before=t.balance_before,
                balance_after=t.balance_after,
                reference_type=t.reference_type,
                reference_id=t.reference_id,
                description=t.description,
                created_at=t.created_at,
            )
            for t in transactions
        ],
        count=len(transactions),
    )


@router.get("/transactions/summary", response_model=UsageSummaryResponse)
async def get_transaction_summary(
    start_date: Optional[datetime] = None,
    end_date: Optional[datetime] = None,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Get transaction summary by type."""
    repo = get_credits_repository(db)
    
    summary = await repo.get_transaction_summary(
        user_id=current_user.id,
        start_date=start_date,
        end_date=end_date,
    )
    
    return UsageSummaryResponse(
        summary=summary,
        period_start=start_date,
        period_end=end_date,
    )


# =============================================================================
# USAGE ENDPOINTS
# =============================================================================

@router.get("/usage/summary", response_model=UsageSummaryResponse)
async def get_usage_summary(
    start_date: Optional[datetime] = None,
    end_date: Optional[datetime] = None,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Get usage summary by type."""
    repo = get_credits_repository(db)
    
    summary = await repo.get_usage_summary(
        user_id=current_user.id,
        start_date=start_date,
        end_date=end_date,
    )
    
    return UsageSummaryResponse(
        summary=summary,
        period_start=start_date,
        period_end=end_date,
    )


@router.get("/usage/daily", response_model=DailyUsageResponse)
async def get_daily_usage(
    days: int = 30,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Get daily usage for the past N days."""
    repo = get_credits_repository(db)
    
    data = await repo.get_daily_usage(
        user_id=current_user.id,
        days=min(days, 90),
    )
    
    return DailyUsageResponse(data=data, days=days)


@router.get("/usage/models", response_model=ModelUsageResponse)
async def get_model_usage(
    days: int = 30,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Get usage statistics by model."""
    repo = get_credits_repository(db)
    
    data = await repo.get_model_usage_stats(
        user_id=current_user.id,
        days=min(days, 90),
    )
    
    return ModelUsageResponse(data=data, days=days)


# =============================================================================
# SUBSCRIPTION PLANS ENDPOINTS
# =============================================================================

@router.get("/plans", response_model=PlansListResponse)
async def get_subscription_plans(
    db: AsyncSession = Depends(get_db),
):
    """Get available subscription plans."""
    repo = get_credits_repository(db)
    
    # Ensure default plans exist
    await repo.create_default_plans()
    await db.commit()
    
    plans = await repo.get_active_plans()
    
    return PlansListResponse(
        plans=[
            PlanResponse(
                id=str(p.id),
                name=p.name,
                display_name=p.display_name,
                description=p.description,
                price_monthly=p.price_monthly,
                price_yearly=p.price_yearly,
                currency=p.currency,
                monthly_credits=p.monthly_credits,
                bonus_credits=p.bonus_credits,
                max_storage_mb=p.max_storage_mb,
                features=p.features or [],
            )
            for p in plans
        ]
    )


# =============================================================================
# ADMIN ENDPOINTS
# =============================================================================

@router.post("/admin/add-credits/{user_id}")
async def admin_add_credits(
    user_id: str,
    data: AddCreditsRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """
    Add credits to a user's balance (admin only).
    """
    if not current_user.is_admin:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Admin access required",
        )
    
    repo = get_credits_repository(db)
    
    transaction = await repo.add_credits(
        user_id=user_id,
        amount=data.amount,
        transaction_type=TransactionType.ADJUSTMENT,
        reference_type="admin_adjustment",
        reference_id=str(current_user.id),
        description=data.reason,
    )
    
    await db.commit()
    
    return {
        "success": True,
        "transaction_id": str(transaction.id),
        "amount": data.amount,
        "user_id": user_id,
    }

File: SmartSpecWeb/server/routers/credits.ts
/**
 * Credit Management tRPC Router
 * Handles credit balance, history, packages, and purchases
 */

import { z } from "zod";
import { router, publicProcedure, protectedProcedure, adminProcedure } from "../_core/trpc";
import {
  getCreditBalance,
  getTransactionHistory,
  getCreditPackages,
  getCreditPackageById,
  addCredits,
  deductCredits,
  getUsageStats,
  type TransactionType,
} from "../services/creditService";

// Zod schemas
const transactionTypeSchema = z.enum([
  "purchase",
  "usage",
  "bonus",
  "refund",
  "adjustment",
  "subscription",
]);

const historyFiltersSchema = z.object({
  limit: z.number().min(1).max(100).default(50),
  offset: z.number().min(0).default(0),
  type: transactionTypeSchema.optional(),
  startDate: z.date().optional(),
  endDate: z.date().optional(),
});

export const creditsRouter = router({
  /**
   * Get current user's credit balance
   */
  balance: protectedProcedure.query(async ({ ctx }) => {
    const balance = await getCreditBalance(ctx.user.id);
    return {
      credits: balance?.credits ?? 0,
      plan: balance?.plan ?? "free",
    };
  }),

  /**
   * Get current user's transaction history
   */
  history: protectedProcedure
    .input(historyFiltersSchema)
    .query(async ({ ctx, input }) => {
      const transactions = await getTransactionHistory({
        userId: ctx.user.id,
        limit: input.limit,
        offset: input.offset,
        type: input.type as TransactionType | undefined,
        startDate: input.startDate,
        endDate: input.endDate,
      });

      return transactions.map((t) => ({
        id: t.id,
        amount: t.amount,
        type: t.type,
        description: t.description,
        balanceAfter: t.balanceAfter,
        createdAt: t.createdAt,
        metadata: t.metadata,
      }));
    }),

  /**
   * Get available credit packages for purchase
   */
  packages: publicProcedure.query(async () => {
    const packages = await getCreditPackages();
    return packages.map((p) => ({
      id: p.id,
      name: p.name,
      description: p.description,
      credits: p.credits,
      priceUsd: parseFloat(p.priceUsd),
      isFeatured: p.isFeatured,
    }));
  }),

  /**
   * Get usage statistics for current user
   */
  stats: protectedProcedure
    .input(z.object({ days: z.number().min(1).max(365).default(30) }))
    .query(async ({ ctx, input }) => {
      return getUsageStats(ctx.user.id, input.days);
    }),

  /**
   * Admin: Add credits to a user
   */
  adminAddCredits: adminProcedure
    .input(
      z.object({
        userId: z.number(),
        amount: z.number().min(1),
        type: transactionTypeSchema,
        description: z.string().min(1).max(512),
        referenceId: z.string().optional(),
      })
    )
    .mutation(async ({ input }) => {
      return addCredits({
        userId: input.userId,
        amount: input.amount,
        type: input.type as TransactionType,
        description: input.description,
        referenceId: input.referenceId,
      });
    }),

  /**
   * Admin: Deduct credits from a user (for adjustments)
   */
  adminDeductCredits: adminProcedure
    .input(
      z.object({
        userId: z.number(),
        amount: z.number().min(1),
        description: z.string().min(1).max(512),
      })
    )
    .mutation(async ({ input }) => {
      return deductCredits({
        userId: input.userId,
        amount: input.amount,
        description: input.description,
        metadata: { reason: "admin_adjustment" },
      });
    }),

  /**
   * Admin: Get any user's credit balance
   */
  adminGetBalance: adminProcedure
    .input(z.object({ userId: z.number() }))
    .query(async ({ input }) => {
      return getCreditBalance(input.userId);
    }),

  /**
   * Admin: Get any user's transaction history
   */
  adminGetHistory: adminProcedure
    .input(
      z.object({
        userId: z.number(),
        ...historyFiltersSchema.shape,
      })
    )
    .query(async ({ input }) => {
      const { userId, ...filters } = input;
      return getTransactionHistory({
        userId,
        ...filters,
        type: filters.type as TransactionType | undefined,
      });
    }),
});
